ibpc_py/src/ibpc/ibpc.py

```diff
@@ -33,21 +37,93 @@ def get_ipd_template(modelname):
+    "ipd_val.zip": "50df37c370557a3cccc11b2e6d5f37f13783159ed29f4886e09c9703c1cad8de",
+    "ipd_test_all.z01": "25ce71feb7d9811db51772e44ebc981d57d9f10c91776707955ab1e616346cb3",
+}
+
 available_datasets = {
-    "ipd": (get_ipd_template("ipd"), ipd_suffixes),
-    "lm": (get_bop_template("lm"), bop_suffixes),
+    "ipd": (get_ipd_template("ipd"), ipd_files),
+    "lm": (get_bop_template("lm"), lm_files),
 }
 
 
+def sha256_file(filename):
+    block_size = 65536
+    sha256 = hashlib.sha256()
+    with open(filename, "rb") as fh:
+        while True:
+            data = fh.read(block_size)
+            if not data:
+                break
+            sha256.update(data)
+    return sha256.hexdigest()
+
+
 def fetch_dataset(dataset, output_path):
-    (url_base, suffixes) = available_datasets[dataset]
-    for suffix in suffixes:
+    (url_base, files) = available_datasets[dataset]
+    fetched_files = []
+    for suffix in files.keys():
+
+        # Sorted so that zip comes before z01
 
         url = url_base + suffix
+
+        outfile = os.path.basename(url)
+        if os.path.exists(outfile):
+            print(f"File {outfile} already present checking hash")
+            computed_hash = sha256_file(outfile)
+            expected_hash = files[suffix]
+            if computed_hash == expected_hash:
+                print(f"File {outfile} detected with expected sha256 skipping download")
+                fetched_files.append(outfile)
+                continue
+            else:
+                print(
+                    f"File {outfile}'s hash {computed_hash} didn't match the expected hash {expected_hash}, downloading again."
+                )
+
         print(f"Downloading from url: {url}")
-        with urlopen(url) as zipurlfile:
-            with ZipFile(BytesIO(zipurlfile.read())) as zfile:
-                zfile.extractall(output_path)
+
+        (filename, headers) = urlretrieve(url, outfile)
+        fetched_files.append(filename)
+
+    for filename in fetched_files:
+        # Append shard if found
+        if filename.endswith("01"):
+            # Let 7z find the other files zipfile can't handle file sharding "multiple disks"
+            fetched_files.remove(filename)
+
+            # Logic for combining files
+            # orig_filename = filename[:-2] + "ip"
+            # combined_filename = "combined_" + orig_filename
+            # with open(combined_filename,'wb') as zipfile:
+            #    with open(orig_filename,'rb') as fd:
+            #        print(f"Appending shard {orig_filename} to {combined_filename}")
+            #        shutil.copyfileobj(fd, zipfile)
+            #    with open(filename,'rb') as fd:
+            #        print(f"Appending shard {filename} to {combined_filename}")
+            #        shutil.copyfileobj(fd, zipfile)
+            # fetched_files.remove(orig_filename)
+            # fetched_files.remove(filename)
+            # fetched_files.append(combined_filename)
```

Comment:
Do we want to leave these comments in?

Reply:
I would really like to figure out how to get rid of the subprocess and leverage ZipFile built in, but so far haven't found out how to reverse the sharding. Hopefully this can be a beacon for others. Though it will probably just not be revisited.

Reply:
Sounds good!